### ISSUE TITLE
SOAP derivatives symmetry issue?

### ISSUE THREAD

**Issue by coolcodecamper** [SINGROUP/dscribe#63]
Dear @lauri-codes,

I am excited to see the new derivatives feature released! Thank you very much for that development!
For some reason, the output differs from my simple numerical implementation.

Below is a small example where dscribe derivatives are compared to a numerical approach using finite differences.
The SOAP descriptor for the HF molecule has got 3+3 features.
The derivative output has been formatted to 6 rows where each row contains the derivatives by x1, y1, z1, x2, y2, z2 of the two atoms.

```
import numpy as np
from ase import Atoms
from ase.vibrations import Vibrations
from dscribe.descriptors import SOAP

atoms = Atoms(symbols='HF', positions=[[7.5, 7.5, 7.5], [7.5, 7.5, 8.4382]], pbc=False)
soap = SOAP(rcut = 6.0, nmax = 1, lmax = 0, species = atoms.get_chemical_symbols())

print('\nSOAP.create:')
print(soap.create(atoms))  # 6 features in total


# DSCRIBE derivatives
dscribe_derivs = soap.derivatives(atoms, method='auto', return_descriptor=False)
n_centers, n_atoms, _, n_features = dscribe_derivs.shape
dscribe_derivs = np.moveaxis(dscribe_derivs, 0, 2).reshape(( 3 * n_atoms, n_atoms * n_features)).T
print('\nSOAP.derivatives:')
print(dscribe_derivs)


# Numerical derivatives using finite displacements
eps = 1e-5  # finite difference
vib = Vibrations(atoms, delta=eps, nfree=2)
atoms_displacements = list(vib.iterimages())  # create finite displacements
s = soap.create(atoms_displacements)
s -= s[0]  # subtract soap for unperturbed structure
s = s[1:]  # pop unperturbed structure
numerical_derivs = 0.5 * (s[1::2] - s[::2]) / eps  # centered finite difference scheme
numerical_derivs = numerical_derivs.reshape((3 * len(atoms), len(s[0].ravel()))).T
print('\nDerivatives by finite displacements:')
print(numerical_derivs)
```

I get the following output:
```
SOAP.create:
[[6.2504344 4.1464286 2.7506678]
 [2.7506678 4.1464286 6.2504344]]

SOAP.derivatives:
[[ 0.         0.         0.         0.         0.         0.       ]
 [ 0.         0.         0.         0.         0.        -3.6276045]
 [ 0.         0.         0.         0.         0.        -4.8129783]
 [ 0.         0.         4.8129783  0.         0.         0.       ]
 [ 0.         0.         3.6276045  0.         0.         0.       ]
 [ 0.         0.         0.         0.         0.         0.       ]]

Derivatives by finite displacements:
[[ 0.         0.         0.         0.         0.         0.       ]
 [ 0.         0.         3.6239624  0.         0.        -3.6239624]
 [ 0.         0.         4.8160553  0.         0.        -4.8160553]
 [ 0.         0.         4.8160553  0.         0.        -4.8160553]
 [ 0.         0.         3.6239624  0.         0.        -3.6239624]
 [ 0.         0.         0.         0.         0.         0.       ]]
```

There are more zeros in the dscribe output. Overall, it looks like a symmetry issue. Which result can be trusted?

**Comment by lauri-codes** [comment]
Hi @coolcodecamper!

Thanks for trying out the new version, any feedback is appreciated!

I think what you're seeing is that in the dscribe output the derivatives of the power spectrum calculated at atom _i_ with respect to the movement of atom _i_ are left unfilled (=zero). I think the motivation was that in most cases you don't need these derivatives, so we don't calculate them to save some time. But whether you need these derivatives or not might depend on your application, so I'm not sure if leaving them unfilled by default is the best choice. I think currently you can force them to be calculated by giving the centers explicitly as cartesian positions, instead of indices or `None` (as in your case).

I guess it has such a minimal impact on performance that we should just always calculate them, or then enable a flag that turns them on or off.

**Comment by coolcodecamper** [comment]
Thank you for the quick reply,
I would highly appreciate this option to compute the full derivatives by default or via flag.

I tried the arguments `positions=[[7.5, 7.5, 7.5], [7.5, 7.5, 8.4382]]`, `positions=[0, 1]` and `include=[0, 1]`. No success.

**Comment by lauri-codes** [comment]
Hi @coolcodecamper,

After taking a more in-depth look, I was wrong about us not calculating the "self-derivatives" (=derivative of SOAP centered on atom _i_ with respect to the movement on atom _i_): we actually do calculate them. There is just a subtle difference in your finite-difference code compared to what is done by dscribe.

In your finite-diffence code, as you displace the atom, you are also displacing the SOAP center that was on top of that atom. This is because you use _indices_ to specify the center locations, as opposed to fixed cartesian positions. dscribe only displaces the atom, not the center that is on top of the atom. This approach keeps the definitions identical irrespective of whether you use indices or cartesian positions as your centers.

If you would like to get the describe behaviour in your finite difference code, you would need to modify the code like this (I also change the output type to 64-bit floating point with the type attribute to get more precise derivative values):

```python
import numpy as np
from ase import Atoms
from ase.vibrations import Vibrations
from dscribe.descriptors import SOAP

positions=[[7.5, 7.5, 7.5], [7.5, 7.5, 8.4382]]
atoms = Atoms(symbols='HF', positions=positions, pbc=False)
soap = SOAP(rcut = 6.0, nmax = 1, lmax = 0, species = atoms.get_chemical_symbols(), dtype="float64")

print('\nSOAP.create:')
print(soap.create(atoms))  # 6 features in total

# DSCRIBE derivatives
dscribe_derivs = soap.derivatives(atoms, method='auto', return_descriptor=False)
n_centers, n_atoms, _, n_features = dscribe_derivs.shape
dscribe_derivs = np.moveaxis(dscribe_derivs, 0, 2).reshape(( 3 * n_atoms, n_atoms * n_features)).T
print('\nSOAP.derivatives:')
print(dscribe_derivs)

# Numerical derivatives using finite displacements
eps = 1e-5  # finite difference
vib = Vibrations(atoms, delta=eps, nfree=2)
atoms_displacements = list(vib.iterimages())  # create finite displacements
s = soap.create(atoms_displacements, [positions]*len(atoms_displacements))
s -= s[0]  # subtract soap for unperturbed structure
s = s[1:]  # pop unperturbed structure
numerical_derivs = 0.5 * (s[1::2] - s[::2]) / eps  # centered finite difference scheme
numerical_derivs = numerical_derivs.reshape((3 * len(atoms), len(s[0].ravel()))).T
print('\nDerivatives by finite displacements:')
print(numerical_derivs)
```
The derivative values in question just happen to be zero in this case since you have specified lmax=0, which means that only the spherically symmetric component is considered, which has a local maximum with respective to moving the atom on top of it. If you increase lmax, you can see that the "self-derivatives" become non-zero.

To me our approach is more consistent, since the SOAP centers are always treated as fixed positions in space. But I'm not sure if there are some merits to the alternative version where you really also move the centers with the atom. Any thoughts on this are very welcome.

**Comment by coolcodecamper** [comment]
Thank you very much for your investigation, @lauri-codes!
After some tests I am certain that I need the derivatives where the SOAP centers are displaced with the atoms (like in my original finite difference code).
It would be great to achieve this behaviour in dscribe (perhaps by some additional flag).

And thank you for pointing out the accuracy and dtype!
Although being quite slow, the finite differences perform much better with float64.

**Comment by lauri-codes** [comment]
Hi @coolcodecamper!

Ok, good to know! I think we can create a flag that controls the behavior. For the numerical derivatives, doing the implementation is very easy. The analytical ones are probably harder and will take more investigation. But this is all transparent to the user if you just go with method="auto".

**Comment by lauri-codes** [comment]
@coolcodecamper: version 1.1.0 now has a new `attach`-argument in the `derivatives`-method that you can use to control this behaviour. If you want to test it out, you can clone the v1.1.0 branch and build it locally. We will also do an official release in the near future.

**Comment by lauri-codes** [comment]
@coolcodecamper The new option is now available in 1.1.0.
